python/thunder/utils/decorators.py

```diff
@@ -0,0 +1,281 @@
+    classes containing model properties to a human and machine parsable format
+    for later analysis or visualization. A few classes under consideration for
+    such decorating include the Transformation class for image alignment and the
+    Source classes for source extraction.
+
+    A key feature of the @serializable decorator is that it can "pickle" data
+    types that are not normally supported by Python's stock JSON dump() and
+    load() methods. Supported datatypes include: list, set, tuple, namedtuple,
+    OrderedDict, datetime objects, numpy ndarrays, and dicts with non-string
+    (but still data) keys. Serialization is performed recursively, and descends
+    into the standard python container types (list, dict, tuple, set).
+
+    IMPORTANT NOTE: The object decorated with @serializable must store their
+    attributes in __slots__ or a __dict__, but not both! For example, you cannot
+    _directly_ wrap a namedtuple in the serializable decorator. However, you can
+    decorate an object that has an attribute pointing at an namedtuple. Only the
+    decorated object is subject to this restriction.
+
+    Some of this code was adapted from these fantastic blog posts by Chris
+    Wagner and Sunil Arora:
+
+      http://robotfantastic.org/serializing-python-data-to-json-some-edge-cases.html
+      http://sunilarora.org/serializable-decorator-for-python-class/
+
+    '''
+
+    class ThunderSerializeableObjectWrapper(object):
+
+        def __init__(self, *args, **kwargs):
+            self.wrapped = cls(*args, **kwargs)
+
+        # Allows transparent access to the attributes of the wrapped class
+        def __getattr__(self, *args):
+            if args[0] != 'wrapped':
+                return getattr(self.wrapped, *args)
+            else:
+                return self.__dict__['wrapped']
+
+        # Allows transparent access to the attributes of the wrapped class
+        def __setattr__(self, *args):
+            if args[0] != 'wrapped':
+                return setattr(self.wrapped, *args)
+            else:
+                self.__dict__['wrapped'] = args[1]
+
+        # Delegate to wrapped class for special python object-->string methods
+        def __str__(self):
+            return self.wrapped.__str__()
+        def __repr__(self):
+            return self.wrapped.__repr__()
+        def __unicode__(self):
+            return self.wrapped.__unicode__()
+
+        # Delegate to wrapped class for special python methods
+        def __call__(self, *args, **kwargs):
+            return self.wrapped.__str__(*args, **kwargs)
+
+        # ------------------------------------------------------------------------------
+        # SERIALIZE()
+
+        def serialize(self, numpyStorage='auto'):
+            '''
+            Serialize this object to a python dictionary that can easily be converted
+            to/from JSON using Python's standard JSON library.
+
+            Arguments
```

Comment:
We've been using a slightly different format for arguments (it's the same format used by `numpy` and `scipy` documentation), should look like this:

```
Parameters
----------
numpyStorage : {'auto', 'ascii', 'base64' }, optional, default 'auto'
    Use to select whether numpy arrays...

Returns
-------
The object encoded as a...
```

Will add this to the style guide!